### DISCUSSION TITLE
slam_toolbox giving error: discarding message because the queue is full

### DISCUSSION THREAD

**Discussion by Illyvia** [iRobotEducation/create3_docs#577] [Questions]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Raspberry Pi 4/ ubuntu 22.04/humble/turtlebot 4 os - VirtualBox ubuntu 22.04 VM/humble 

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.2.6

### Which RMW is your robot running?

FastRTPS

### Is the robot connected to a network? If so, what is the network type?

home network

### Are there multiple Create 3 robots connected to your network?

None

### Is multicast enabled?

None

### What is the Adapter Board's USB/BLE Toggle currently switched to?

USB

### Describe your question.

I have recently purchased a SLAMTEC RPLIDAR A1, and was hoping to use it with SLAM, and python. Testing the create3 examples with SLAM, I get the repeated error from slam_toolbox:  `[async_slam_toolbox_node-1] [INFO] [1711769129.930220100] [slam_toolbox]: Message Filter dropping message: frame 'laser_frame' at time 1711769129.807 for reason 'discarding message because the queue is full'
[async_slam_toolbox_node-1] [INFO] [1711769129.942137353] [slam_toolbox]: Message Filter dropping message: frame 'rplidar_link' at time 1711769129.929 for reason 'discarding message because the queue is full'`

I tried this with the create3 examples, and with the turtlebot4 ones; same error on both.

**Comment by alsora** [comment]
This problem seems to be caused by the tf topic.

You should check whether your transformations tree is complete.
The error message shows that you have both a `rplidar_link` and a `laser_frame` frames, I'm not sure if this is expected.

Can you show how are you running slam toolbox and the lidar?

If the transformations are good, another possible cause could be that the timestamps of the robot and the Pi are not synchronized.
You can check that by comparing the timestamp field of tf messages published by the Raspberry Pi.
This scenario can be fixed by setting up an NTP server on the Pi as described here https://iroboteducation.github.io/create3_docs/setup/compute-ntp/

**Reply by Illyvia** [reply]
The NTP server is preinstalled on turtlebot4 image. I have checked with: `sudo chronyc clients`. 

As for showing how I'm running slam toolbox and the lidar, I just cloned the create3 examples, built it, and ran the commands from it:

`ubuntu@ubuntu:~/create3_examples_ws$ ros2 launch create3_lidar_slam sensors_launch.py
[INFO] [launch]: All log files can be found below /home/ubuntu/.ros/log/2024-03-30-03-42-28-080948-ubuntu-74066
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [static_transform_publisher-1]: process started with pid [74105]
[static_transform_publisher-1] [WARN] [1711770148.716080596] []: Old-style arguments are deprecated; see --help for new-style arguments
[static_transform_publisher-1] [INFO] [1711770148.864449307] [static_transform_publisher_cOmevmprsK4idQif]: Spinning until stopped - publishing transform
[static_transform_publisher-1] translation: ('-0.012000', '0.000000', '0.144000')
[static_transform_publisher-1] rotation: ('0.000000', '0.000000', '0.000000', '1.000000')
[static_transform_publisher-1] from 'base_footprint' to 'laser_frame'
[INFO] [rplidar_composition-2]: process started with pid [74141]
[rplidar_composition-2] [INFO] [1711770150.917096106] [rplidar_node]: RPLIDAR running on ROS 2 package rplidar_ros. SDK Version: '1.12.0'
[rplidar_composition-2] [INFO] [1711770153.425100219] [rplidar_node]: RPLIDAR S/N: BE8CED95C4E493CBA5E69EF02F5C4B6F
[rplidar_composition-2] [INFO] [1711770153.425238810] [rplidar_node]: Firmware Ver: 1.29
[rplidar_composition-2] [INFO] [1711770153.425282458] [rplidar_node]: Hardware Rev: 7
[rplidar_composition-2] [INFO] [1711770153.426758793] [rplidar_node]: RPLidar health status : '0'
[rplidar_composition-2] [INFO] [1711770153.444181561] [rplidar_node]: Start
[rplidar_composition-2] [INFO] [1711770153.988886950] [rplidar_node]: current scan mode: Sensitivity, max_distance: 12.0 m, Point number: 7.9K , angle_compensate: 2, flip_x_axis 0`

And for the slam toolbox: 

`ubuntu@ubuntu:~$ ros2 launch create3_lidar_slam slam_toolbox_launch.py
[INFO] [launch]: All log files can be found below /home/ubuntu/.ros/log/2024-03-30-03-42-41-613827-ubuntu-74237
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [sync_slam_toolbox_node-1]: process started with pid [74258]
[sync_slam_toolbox_node-1] [INFO] [1711770162.916572180] [slam_toolbox]: Node using stack size 40000000
[sync_slam_toolbox_node-1] [INFO] [1711770163.237321037] [slam_toolbox]: Using solver plugin solver_plugins::CeresSolver
[sync_slam_toolbox_node-1] [INFO] [1711770163.240566686] [slam_toolbox]: CeresSolver: Using SCHUR_JACOBI preconditioner.
[sync_slam_toolbox_node-1] [INFO] [1711770163.351582444] [slam_toolbox]: Message Filter dropping message: frame 'rplidar_link' at time 1711770163.228 for reason 'discarding message because the queue is full'
[sync_slam_toolbox_node-1] [INFO] [1711770163.462025746] [slam_toolbox]: Message Filter dropping message: frame 'rplidar_link' at time 1711770163.351 for reason 'discarding message because the queue is full'`

It then repeats the queue is full message indefinitely.

On the VM: `ros2 launch create3_lidar_slam rviz_launch.py`

`root@ROS4: /home/ros4# ros2 launch create3_lidar_slam rviz_launch.py
[INFO] [launch]: All log files can be found below /root/.ros/log/2024-03-30-08-38-32-023302-ROS4-2965
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [rviz2-1]: process started with pid [2966]
[rviz2-1] Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
[rviz2-1] QStandardPaths: runtime directory '/run/user/1000' is not owned by UID 0, but a directory permissions 0700 owned by UID 10
00 GID 1000
[rviz2-1] QStandardPaths: runtime directory '/run/user/1000' is not owned by UID 0, but a directory permissions 0700 owned by UID 10
00 GID 1000
[rviz2-1] [INFO] [1711805913.542830833] [rviz2]: Stereo is NOT SUPPORTED
[rviz2-1] [INFO] [1711805913. 550821535][rviz2]: OpenGl version: 4.5 (GLSL 4.5)
[rviz2-1] [INFO] [1711805913.652430682] [rviz2]: Stereo is NOT SUPPORTED
[rviz2-1] [INFO] [1711805916.380035888] [rviz]: Message Filter dropping message: frame 'rplidar_link' at time 1711805914.238 for rea
son 'discarding message because the queue is full'
[rviz2-1] [INFO] [1711805916.582240359] [rviz]: Message Filter dropping message: frame 'rplidar_link' at time 1711805914.410 for rea
son 'discarding message because the queue is full'
[rviz2-1] [INFO] [1711805916. 780601862] [rviz]: Message Filter dropping message: frame 'rplidar_link' at time 1711805914.508 for rea
son 'discarding message because the queue is full'
[rviz2-1] [INFO] [1711805916.981441678] [rviz]: Message Filter dropping message: frame 'rplidar_link' at time 1711805914.667 for rea
son 'discarding message because the queue is full'
[rviz2-1] [INFO] [1711805917. 179681749] [rviz]: Message Filter dropping message: frame 'rplidar_link' at time 1711805914.791 for rea
son 'discarding message because the queue is full'
[rviz2-1] [INFO] [1711805917 . 379945254] [rviz]: Message Filter dropping message: frame 'rplidar_link' at time 1711805914.938 for rea
son 'discarding message because the queue is full'
[rviz2-1] [INFO] [1711805917.580434997] [rviz]: Message Filter dropping message: frame 'rplidar_link' at time 1711805915.073 for rea
son 'discarding message because the queue is full'
[rviz2-1] [INFO] [1711805917.780280987] [rviz]: Message Filter dropping message: frame 'rplidar_link' at time 1711805915.355 for rea
son 'discarding message because the queue is full'
[rviz2-1] [INFO] [1711805917.980611374] [rviz]: Message Filter dropping message: frame 'rplidar_link' at time 1711805915.650 for rea
son 'discarding message because the queue is full'
[rviz2-1] [INFO] [1711805918.179753535] [rviz]: Message Filter dropping message: frame 'laser_frame' at time 1711805915.908 for reas
on 'discarding message because the queue is full'`

![image](https://github.com/iRobotEducation/create3_docs/assets/122619961/284e3998-8f0b-45ed-82ed-806a9dc4f984)

**Reply by alsora** [reply]
Can you show the available tfs?
Running this while everything else is up and running
```
ros2 run tf2_tools view_frames
```
This will generate a pdf that you can upload here

See details here https://docs.ros.org/en/humble/Tutorials/Intermediate/Tf2/Introduction-To-Tf2.html#using-view-frames

**Reply by Illyvia** [reply]
[frames_2024-03-30_10.15.30.pdf](https://github.com/iRobotEducation/create3_docs/files/14812809/frames_2024-03-30_10.15.30.pdf)

**Reply by alsora** [reply]
The `laser_frame` is attached to `base_footprint` which doesn't have any parent.
At the very least those should be attached to the `base_link`.

Moreover, all tfs have a 0 timestamp.

**Reply by Illyvia** [reply]
Is there anything I can do to fix this? (Sorry, I'm really new to ROS.)

**Reply by brianabouchard** [reply]
@alsora the Create 3 Examples repository has `laser_frame` connected to `base_footprint`, but I've never been clear on why that is - it was that way before we added the most recent updates to that example. I think the big missing link here is that there is no `odom` and no `map.` `odom` should be the parent of both `base_footprint` and `base_link`

@TheInventor2023 here's a link to a sample tf tree for what it should look like: https://github.com/iRobotEducation/create3_examples/blob/humble/create3_lidar_slam/TF_Tree.pdf

As for why this is happening and how to fix it, I'm not 100% sure. Whenever I've had similar issues, it is almost always a time sync issue. Do the Create 3 logs (from the webserver) have any messages that indicate it is unable to complete the time sync? Usually says something about the delay being too high?

I would also suggest trying the custom ntp config outlined here: https://github.com/turtlebot/turtlebot4/issues/216. That made a big difference for me.

**Reply by alsora** [reply]
We use `base_footprint` rather than `base_link` because the latter can oscillate due to small variations in roll and pitch angles reported by the IMU and this causes some problems to `slam_toolbox`.

And yes, this should be somehow connected to `odom`.
@brianabouchard I thought that it should have been "odom" -> "base_link" -> "base_footprint" -> "laser_fram", but your picture is likely correct.
The transformation "odom" -> "base_footprint" should be available regarldess of `slam_toolbox`.

The lack of `map` is likely because `slam_toolbox` hasn't been able to process anything yet, so it hasn't published this frame.

I was going to do some tests on Monday on my Create 3 and report back, as I don't remember much of this out of the top of my head.

**Reply by Illyvia** [reply]
Hmm, I tried adding the unicast xml file, and it changed some things; `ubuntu@ubuntu:~$ ros2 launch create3_lidar_slam sensors_launch.py
[INFO] [launch]: All log files can be found below /home/ubuntu/.ros/log/2024-03-31-18-35-53-739963-ubuntu-2946
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [static_transform_publisher-1]: process started with pid [2947]
[static_transform_publisher-1] [WARN] [1711910154.410235991] []: Old-style arguments are deprecated; see --help for new-style arguments
[static_transform_publisher-1] [INFO] [1711910154.476371586] [static_transform_publisher_quPva3VxsABjlTQc]: Spinning until stopped - publishing transform
[static_transform_publisher-1] translation: ('-0.012000', '0.000000', '0.144000')
[static_transform_publisher-1] rotation: ('0.000000', '0.000000', '0.000000', '1.000000')
[static_transform_publisher-1] from 'base_footprint' to 'laser_frame'
[INFO] [rplidar_composition-2]: process started with pid [2957]
[rplidar_composition-2] [INFO] [1711910156.479086478] [rplidar_node]: RPLIDAR running on ROS 2 package rplidar_ros. SDK Version: '1.12.0'
[rplidar_composition-2] [INFO] [1711910158.987768510] [rplidar_node]: RPLIDAR S/N: BE8CED95C4E493CBA5E69EF02F5C4B6F
[rplidar_composition-2] [INFO] [1711910158.987981878] [rplidar_node]: Firmware Ver: 1.29
[rplidar_composition-2] [INFO] [1711910158.988144803] [rplidar_node]: Hardware Rev: 7
[rplidar_composition-2] [INFO] [1711910158.989809494] [rplidar_node]: RPLidar health status : '0'`

It only goes up to that point now, and the LIDAR stops spinning.

As for slam toolbox, it stays pretty much the same:

`ubuntu@ubuntu:~$ ros2 launch create3_lidar_slam slam_toolbox_launch.py
[INFO] [launch]: All log files can be found below /home/ubuntu/.ros/log/2024-03-31-18-36-01-673638-ubuntu-2968
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [async_slam_toolbox_node-1]: process started with pid [2969]
[async_slam_toolbox_node-1] [INFO] [1711910162.302568670] [slam_toolbox]: Node using stack size 40000000
[async_slam_toolbox_node-1] [INFO] [1711910162.727194520] [slam_toolbox]: Using solver plugin solver_plugins::CeresSolver
[async_slam_toolbox_node-1] [INFO] [1711910162.729349021] [slam_toolbox]: CeresSolver: Using SCHUR_JACOBI preconditioner.`

I also tried the custom ntp, but it didn't really work: `Mar 31 17:56:01 iRobot-060EB663C12A45E7BC6AE2340FB9830F user.notice ntpd: ntpd: reply from 192.168.186.3: delay 0.006247 is too high, ignoring
Mar 31 18:00:32 iRobot-060EB663C12A45E7BC6AE2340FB9830F user.notice ntpd: Script called in periodic mode
Mar 31 18:06:50 iRobot-060EB663C12A45E7BC6AE2340FB9830F user.notice ntpd: ntpd: reply from 192.168.186.3: delay 0.003295 is too high, ignoring`

I'm going to see if factory resetting my robot, and doing a fresh install of ubuntu could make any difference.

**Reply by Illyvia** [reply]
Uhh, this is weird; after a clean install of ubuntu, and a factory reset, it seems to work!
![Screenshot 2024-03-31 171109](https://github.com/iRobotEducation/create3_docs/assets/122619961/e6245c50-473d-4e88-b9ca-91379d301af4)

**Reply by brianabouchard** [reply]
Well yay! I'm glad it is solved, but odd that there wasn't a clear reason.

**Reply by Illyvia** [reply]
Sorry to reopen the problem, but upon running rviz again, it gives the same error, and the map doesn't show up. I'm not sure why that one time it worked.
